api/prometheus/v1/api.go

```diff
@@ -17,17 +17,83 @@ package v1
 
 import (
 	"context"
-	"encoding/json"
 	"errors"
 	"fmt"
 	"net/http"
 	"strconv"
 	"time"
+	"unsafe"
+
+	json "github.com/json-iterator/go"
 
 	"github.com/prometheus/client_golang/api"
 	"github.com/prometheus/common/model"
 )
 
+func init() {
+	json.RegisterTypeEncoderFunc("model.SamplePair", marshalPointJSON, marshalPointJSONIsEmpty)
+	json.RegisterTypeDecoderFunc("model.SamplePair", unMarshalPointJSON)
+}
+
+func unMarshalPointJSON(ptr unsafe.Pointer, iter *json.Iterator) {
+	p := *((*model.SamplePair)(ptr))
+	if !iter.ReadArray() {
+		iter.ReportError("unmarshal model.SamplePair", "value must be an array")
+		return
+	}
+	p.Timestamp = model.Time(iter.ReadFloat64())
+	if !iter.ReadArray() {
+		iter.ReportError("unmarshal model.SamplePair", "array must be of len 2")
```

Comment:
The comment is a bit unclear, but do we even need this check?

Shouldn't `f, err := strconv.ParseFloat(iter.ReadString(), 64)` just fail if there is nothing else to read?

Reply:
The iter requires calling readarray to proceed to the next item in the array, once you call it you have to call it until it returns false or you set an error.

Reply:
yes I missed that part, thanks.
Maybe something like this would be a bit more clear for someone like me who is not that familiar with this package:

```
// Move the iterator to the the sample value.
if !iter.ReadArray() {
		iter.ReportError("unmarshal model.SamplePair", "array doesn't include a sample value")
```
can be even without the extra comment.

Reply:
I changed the ReadArray errors to all be the same error `"SamplePair must be [timestamp, value]"` which is probably lots more descriptive anyways.